AWS EBS volume support

justinsb opened the pull request:
Code definitely doesn't work yet, and an open discussion in #5129 on whether we should write server & client code once (with a provider approach) or N times (with a type-per-cloud approach).

But would greatly appreciate any comments particularly on the pkg/cloudprovider/aws/aws.go code.  The extra code is non-trivial because AWS volume mounting is pretty painful: you have to specify an unused mount device.

justinsb committed:
Add initial support for Volumes to AWS

justinsb commented:
AWS code is mostly here: https://github.com/GoogleCloudPlatform/kubernetes/commit/bd9021e94f026791c563613c5c2832ce8f78f235

markturansky commented on pkg/api/types.go:
```diff
@@ -173,9 +173,9 @@ type VolumeSource struct {
+	// CloudPersistentDisk represents a cloud-provided storage resource that is attached to a
 	// kubelet's host machine and then exposed to the pod.
-	GCEPersistentDisk *GCEPersistentDiskVolumeSource `json:"persistentDisk"`
+	CloudPersistentDisk *CloudPersistentDiskVolumeSource `json:"persistentDisk"`
```
I don't think this is going to work.  You'll overload this struct for every type of cloud provider, which means you have to type it somehow.  Some properties would be relevant to some but not all of the providers.  

Better, I think, to have an explicit struct per type.

We agree there will be commonalities between them.  Good code structure can promote reuse for the common pieces.

justinsb replied:
Yeah, I'd started this before I saw that there were different viewpoints on #5129.  Really looking for feedback on the AWS stuff specifically; once we figure out #5129 then I'll rebase the AWS work on to whatever we choose.

swagiaal replied:
@justinsb could you modify this so it implements AWS persistent volumes as sibling to GCE PD. Similar to other proposed persistent volume plugins: #4612, and #4601.
I think there is general consensus on doing that for now, then improving the abstraction to promote code sharing after a couple of them are merged. See the discussion here on  https://github.com/GoogleCloudPlatform/kubernetes/pull/4601/files#r25993686

markturansky commented:
cc @swagiaal

justinsb committed:
Support multiple k8s clusters

justinsb committed:
Fix tests

justinsb committed:
Fix AWS region vs zone

We were specifying a region, but naming it as a zone in util.sh

The zone matters just as much as the region, e.g. for EBS volumes.

We also change the config to require a Zone, not a Region.
But we fallback to get the information from the metadata service.

justinsb committed:
Always create volumes in the active k8s zone

justinsb committed:
Simplify logic of pd.go

piosz commented:
Is this PR active? If not please close it according to https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/pull-requests.md

justinsb commented:
I should be ready with the updated code this week.  I'm currently working on load balancer support, and maintaining two big feature branches simultaneously is too hard, so I'm till that merges before working on EBS.

swagiaal commented:
@justinsb I can work on this if you are too busy with other PRs. It was on my radar before you posted this.

justinsb commented:
I think it's really close, and not sure if it makes any sense to attempt a hand-off.

Let me try to get some of my smaller PRs merged, so I only have to deal with this one and the ELB one...  That should be manageable.

swagiaal commented:
@justinsb Sounds good :)

justinsb committed:
Fix merge problems

justinsb committed:
Fix tests

justinsb committed:
Strip leading slash from url path when parsing aws volume

justinsb committed:
Update IAM permissions for minion, to allow EBS

justinsb committed:
Fix mountpoint values

justinsb committed:
Tolerate volume being already-attached

justinsb committed:
Fix comment

justinsb committed:
Small clean-ups

justinsb committed:
Add safe_format_and_mount script to aws

Apache licensed :-)

justinsb committed:
Add missing import

justinsb committed:
Create the /usr/share/google dir in salt

justinsb committed:
More logging around error causes

Come back exceptions, all is forgiven!

justinsb committed:
Parse the pdName from the volume mount

Don't assume there are no slashes!

justinsb committed:
Default attachment status to detached

So no-attachments = status detached

justinsb committed:
Fix pd name parse

justinsb committed:
Fix detached-check logic, warn on multiple attachments

justinsb committed:
Add comment about EBS status being a bit slow through API

justinsb commented:
This now "works for me", but I need to run the e2e tests (which I have extended to cover AWS).

Should I squash?   I don't know if anyone really cares to see my thought process here :-)  And there hasn't been that much review activity yet...

markturansky commented on pkg/api/types.go:
```diff
@@ -189,6 +189,9 @@ type VolumeSource struct {
 	GCEPersistentDisk *GCEPersistentDiskVolumeSource `json:"gcePersistentDisk"`
+	// AWSPersistentDisk represents a AWS EBS disk that is attached to a
+	// kubelet's host machine and then exposed to the pod.
+	AWSPersistentDisk *AWSPersistentDiskVolumeSource `json:"awsPersistentDisk"`
```
Would AWSElasticBlockStore be more accurately named?

justinsb replied:
Perhaps.  EBS is just a marketing term though, right?  I'd rather think about the API in terms of what the user is trying to achieve - they want a persistent volume.  Maybe it should be AWSPersistentVolume.

I guess the question is when Amazon announces ElasticNvram, which is still a block device, would we create a second entry in the API?  (Or, if they had chosen to launch SSD under a different name).  I'd argue we shouldn't, and as such we shouldn't mirror Amazon's trademark-du-jour...

markturansky replied:
I believe @bgrant0607's intention is to remove all these types of disk from VolumeSource and from the end user (pod authors and the like).  The user will make a request for a persistent volume (via PVClaim) and get something.  They won't know what and they shouldn't care.

The cluster admin, OTOH, deals with specific infrastructure.  He needs to create a specific kind of disk.  ```AWSElasticBlockStore``` maps cleanly to the AWS API type EbsBlockDevice.

justinsb replied:
OK, I'll change it.  Avoiding another rebase is more important to me than nomenclature :-)

markturansky commented on pkg/api/types.go:
```diff
@@ -367,6 +373,28 @@ type GCEPersistentDiskVolumeSource struct {
+// A AWS EBS disk can only be mounted as read/write once.
+type AWSPersistentDiskVolumeSource struct {
+	// Unique name of the PD resource. Used to identify the disk in AWS
+	PDName string `json:"pdName"`
```
Similarly, the AWS API has "volumeId", not PDName.

I see why you were thinking of 1 "cloud disk" for re-use across both GCE and AWS.  But so long as they must be separate implementations, I think the names of fields can map cleanly and obviously to the AWS API as opposed to mapping to the GCE API.  My $0.02.

justinsb replied:
Fair enough!  Making the change now.  VolumeId feels like a better name even for GCE :-)

justinsb replied:
Also, this isn't just the AWS ID.  We qualify it with the AZ e.g. aws://us-east-1a/vol-12345678

markturansky commented on pkg/volume/aws_pd/aws_pd.go:
```diff
@@ -0,0 +1,320 @@
+var _ volume.VolumePlugin = &awsPersistentDiskPlugin{}
+
+const (
+	awsPersistentDiskPluginName = "kubernetes.io/aws-pd"
```
aws-ebs?

justinsb replied:
My inclination is to keep this named-by-intent, not named by product-name.

markturansky replied:
Perhaps this is changed to match what consensus comes regarding the name in VolumeSource?

"aws-pd", to me, requires parsing and thought while "aws-ebs" is known AWS infrastructure.

markturansky commented:
Also see fuzzer.go for testing the new volume:   https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/testing/fuzzer.go#L176

justinsb committed:
Rename pdName -> volumeId for AWS persistent volumes

justinsb committed:
Fixup merge mistakes

justinsb committed:
Add AWSPersistentDisk to fuzzer

justinsb commented:
Added fuzzer, rebased to avoid conflicts with iSCSI

justinsb committed:
Rename AWSPersistentDisk -> AWSElasticBlockStore, aws-pd -> aws-ebs

Per comments from @markturansky - thanks!

justinsb committed:
Apply latest changes from copy-and-pasted gce_pd

justinsb committed:
Add missing conversion for v1beta2

justinsb committed:
Make fetching the aws instance id optional, so we can use it on e2e

justinsb committed:
Remove now-unused instanceId parameter from newAwsCloud

justinsb committed:
Provide more output during a disk delete

justinsb commented:
Now with passing e2e tests also!

http://public-kubernetes-meteor.s3-website-us-east-1.amazonaws.com/builds/kubernetes-e2e-aws/6/_1.xml

In particular:

```
<testcase name="PD should schedule a pod w/ a RW PD, remove it, then schedule it on another host" classname="Kubernetes e2e Suite run 1 of 1" time="60.902563537"/>
```

markturansky commented on cmd/e2e/e2e.go:
```diff
@@ -63,5 +68,22 @@ func main() {
+	if context.Provider == "aws" {
+		awsConfig := "[Global]\n"
+		if cloudConfig.Zone == "" {
+			glog.Error("gce_zone must be specified for AWS")
```
is "gce_zone" correct here?

justinsb replied:
This is tricky.  I'm not sure whether we want flags for each provider (i.e. potentially duplicate kube_master, gce_project, gce_zone -> aws_kube_master, aws_gce_project, aws_gce_zone) or whether we should treat these as "cloud_project" & "cloud_zone".  If the latter, a secondary question is whether we should rename the flags.

markturansky replied:
You can only have one cloud provider at a time, right?  Perhaps this is a good place for generic cloud patterns, so long as it's not planned to allow a cluster span infrastructure providers...   If that's the case, that's probably an easy small PR instead of adding more to this one.

thockin replied:
As long as the concepts fundamentally match up, I'm in favor of generification.

thockin replied:
I'm fine with not renaming in this PR, but please file a bug on this.  We just need to figure out the ripple of renaming them.

markturansky commented on pkg/api/types.go:
```diff
@@ -208,6 +211,9 @@ type PersistentVolumeSource struct {
 	// GCEPersistentDisk represents a GCE Disk resource that is attached to a
 	// kubelet's host machine and then exposed to the pod.
 	GCEPersistentDisk *GCEPersistentDiskVolumeSource `json:"gcePersistentDisk"`
+	// AWSElasticBlockStore represents a AWS EBS disk that is attached to a
```
s/a/an

justinsb replied:
Fixed - thanks!

markturansky commented:
LGTM for what I can review.  You'll need @thockin for the rest.

justinsb committed:
Grammar fix: s/a AWS/an AWS/g

thockin commented on pkg/api/types.go:
```diff
@@ -189,6 +189,9 @@ type VolumeSource struct {
 	GCEPersistentDisk *GCEPersistentDiskVolumeSource `json:"gcePersistentDisk"`
+	// AWSElasticBlockStore represents an AWS EBS disk that is attached to a
+	// kubelet's host machine and then exposed to the pod.
+	AWSElasticBlockStore *AWSElasticBlockStoreVolumeSource `json:"awsElasticBlockStore"`
```
@bgrant0607 Are you OK with AWS support here to parallel GCE?  It seems appropriate.  We can EOL them both when persistent volumes are up to snuff.

thockin commented on pkg/cloudprovider/aws/aws.go:
```diff
@@ -506,3 +573,383 @@ func (self *AWSCloud) GetZone() (cloudprovider.Zone, error) {
+}
+
+// TODO: Also return number of mounts allowed?
+func (self *awsInstanceType) getEbsMountDevices() []string {
```
Here and elsewhere: Go style would say "EBS"

thockin commented on pkg/cloudprovider/aws/aws.go:
```diff
@@ -506,3 +573,383 @@ func (self *AWSCloud) GetZone() (cloudprovider.Zone, error) {
+	ec2 EC2
+
+	// id in AWS
+	awsId string
```
nit: ID

thockin commented on pkg/cloudprovider/aws/aws.go:
```diff
@@ -506,3 +573,383 @@ func (self *AWSCloud) GetZone() (cloudprovider.Zone, error) {
+
+// Implements Volumes.AttachDisk
+func (aws *AWSCloud) AttachDisk(instanceName string, diskName string, readOnly bool) (string, error) {
+	disk, err := newAwsDisk(aws.ec2, diskName)
```
AWS  vs Aws

thockin commented on pkg/volume/aws_pd/aws_pd.go:
```diff
@@ -0,0 +1,320 @@
+var _ volume.VolumePlugin = &awsElasticBlockStorePlugin{}
+
+const (
+	awsElasticBlockStorePluginName = "kubernetes.io/aws-ebs"
```
This whole pkg and the files should be called aws_ebs

thockin commented:
LGTM but for naming stuff

justinsb committed:
Style: Ebs -> EBS

justinsb committed:
Style: awsId -> awsID

justinsb committed:
Style: Aws -> AWS

justinsb committed:
Style: volumeId -> volumeID

justinsb committed:
Rename aws_pd -> aws_ebs

justinsb committed:
Fix some mistaken volumeId -> volumeID changes

justinsb commented:
So I think we're mainly waiting on @bgrant0607 to comment on the API (?)  https://github.com/GoogleCloudPlatform/kubernetes/pull/5138#discussion_r28034332

And I'm not entirely sure what we want to do about the parameter names in e2e.go, at least in this PR?  (https://github.com/GoogleCloudPlatform/kubernetes/pull/5138#discussion_r28027669)

thockin commented:
Don't wait too long for Brian, I am pretty sure it's a case of "<sigh>, ok"
:)



On Fri, Apr 10, 2015 at 1:03 PM, Justin Santa Barbara <
notifications@github.com> wrote:

> So I think we're mainly waiting on @bgrant0607
> <https://github.com/bgrant0607> to comment on the API (?) #5138 (comment)
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5138#discussion_r28034332>
>
> And I'm not entirely sure what we want to do about the parameter names in
> e2e.go, at least in this PR? (#5138 (comment)
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5138#discussion_r28027669>
> )
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5138#issuecomment-91667200>
> .
>

thockin commented on pkg/volume/aws_ebs/aws_pd.go:
```diff
@@ -0,0 +1,320 @@
+limitations under the License.
+*/
+
+package aws_ebs
```
nit: file should be renamed to match dir

thockin commented on pkg/volume/aws_ebs/aws_pd_test.go:
```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
```
Same rename comment - file name.

thockin commented:
Needs a rebase too, otherwise LGTM

justinsb committed:
Rename aws_pd.go  -> aws_ebs.go, aws_pd_test.go -> aws_ebs_test.go

justinsb commented:
Opened issue #6699 to track renaming the e2e cmd arguments.

lavalamp commented:
@thockin, @justinsb -- this definitely could have used a squash

thockin commented:
Yeah, I messed up - meant to ask for it, forgot.

I wonder if there are tools that could help here?


On Fri, Apr 10, 2015 at 3:48 PM, Daniel Smith <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin>, @justinsb
> <https://github.com/justinsb> -- this definitely could have used a squash
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5138#issuecomment-91711145>
> .
>

justinsb commented:
Sorry, I should definitely have squashed it once we were getting closer.  But squashing also obliterates history.  I have looked, and haven't found any great solution.  Even with squashing, I still don't love the merge-commits that the button produces.

I hear code.google.com will soon be available.  Let's build v2 (on gerrit this time!)

Or we could probably figure out a script that squashes the commits, changes the author and does a push.

bgrant0607 commented:
Sorry for the late response. Github notifications are useless, since I get hundreds per day.

Fortunately, I'm ok with this. Next time, please ask someone to ping me not through github for an API change.

delianides commented:
Is there going to be any documentation for this? How exactly does this work with availability zones? Should the volume exists in the same AZ as the master?

tslater commented:
I would love to see some documentation as well.
